### PR TITLE
Add `nnx.training`

### PR DESCRIPTION
**Pull request by chiamp**
Add `nnx.training`.

API reference for [`nnx.metrics`](https://flax--3782.org.readthedocs.build/en/3782/api_reference/flax.experimental.nnx/training/metrics.html) and [`nnx.optimizer`](https://flax--3782.org.readthedocs.build/en/3782/api_reference/flax.experimental.nnx/training/optimizer.html).

~Had to import [`chex.ArrayTree`](https://github.com/google/flax/pull/3782/files#diff-7bb758349c08d151354150d8d215d8b98a0ce1308134088bafc41c395c59074aR126) to make typing happy. Is there an alternative where we don't have to use this dependency? e.g. annotating with `# type: ignore`.~

Renamed `optimizer.update` method to `optimizer.apply_gradients` method, because `optimizer.update` was being called instead of the base class `nnx.Module.update` method when using `nnx.jit`.
Added `OptState` and `MetricState` `Variable` wrapper classes for `nnx.optimizer.Optimizer` and `nnx.metrics.Metric`, respectively.

Updated [`toy_examples/02_lifted_transforms.py`](https://github.com/chiamp/flax/blob/nnx_module_subclass/flax/experimental/nnx/examples/toy_examples/02_lifted_transforms.py) to use `nnx.optimizer.Optimizer`. Training loop runs slightly slower.
Updated [`docs/experimental/nnx/mnist_tutorial.ipynb`](https://flax--3782.org.readthedocs.build/en/3782/experimental/nnx/mnist_tutorial.html) to use `nnx.metrics.MultiMetric` and `nnx.optimizer.Optimizer`.

For some reason, I'm getting an `nbstripout` error on CI when I try to include the outputs of the mnist tutorial, so I have removed them for now.

### CODE REVIEW

File: docs/experimental/nnx/mnist_tutorial.md [hunk 1]
```diff
@@ -170,18 +174,23 @@ This function takes the `state` and a data `batch` and does the following:
 
 ```{code-cell} ipython3
 @jax.jit
-def train_step(state: TrainState, batch):
+def train_step(state: nnx.State, static: nnx.GraphDef, batch):
```
Comment (cgarciae) [inline docs/experimental/nnx/mnist_tutorial.md:177]: maybe we could use `nnx.jit` here to make it simpler?

Comment (chiamp) [reply docs/experimental/nnx/mnist_tutorial.md]: Done.

Should we show some examples elsewhere on how to use NNX with both `nnx.jit` and `jax.jit`; in toy examples, and/or a guide with a side-by-side comparison, similar to the upgrade guides?

**Review by cgarciae — APPROVED** [review]
